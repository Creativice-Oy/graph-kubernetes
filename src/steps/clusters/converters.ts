import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Cluster } from '@kubernetes/client-node';
import { Entities } from '../constants';

function getClusterKey(id: string) {
  return `kube_cluster:${id}`;
}

export function createClusterEntity(data: Cluster) {
  const { caData, caFile, ...clusterData } = data;

  return createIntegrationEntity({
    entityData: {
      source: {
        ...clusterData,
      },
      assign: {
        _class: Entities.CLUSTER._class,
        _type: Entities.CLUSTER._type,
        _key: getClusterKey(data.name),
        name: data.name,
        displayName: data.name,
        server: data.server,
        skipTlsVerify: data.skipTLSVerify,
      },
    },
  });
}
