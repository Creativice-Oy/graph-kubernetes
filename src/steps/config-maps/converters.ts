import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createConfigMapEntity(configMap: k8s.V1ConfigMap) {
  /*
    ConfigMap contains 2 fields - data and binary data.
    They expose the actual data that's stored inside the map.
    We probably don't want to store this in _rawData
      1) Could be big (?)
      2) User may store something sensitive inside (even though that'd be against ConfigMap recommended use)
  */
  const { binaryData, data, ...withoutData } = configMap;

  return createIntegrationEntity({
    entityData: {
      source: withoutData,
      assign: {
        _class: Entities.CONFIGMAP._class,
        _type: Entities.CONFIGMAP._type,
        _key: withoutData.metadata?.uid,
        name: withoutData.metadata?.name,
        displayName: withoutData.metadata?.name,
        generation: withoutData.metadata?.generation,
        deletionGracePeriodSeconds:
          withoutData.metadata?.deletionGracePeriodSeconds,
        resourceVersion: withoutData.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(
          withoutData.metadata?.creationTimestamp,
        ),
        deletedOn: parseTimePropertyValue(
          withoutData.metadata?.deletionTimestamp,
        ),
        // spec fields
        immutable: withoutData.immutable,
        classification: null,
        encrypted: false,
      },
    },
  });
}
