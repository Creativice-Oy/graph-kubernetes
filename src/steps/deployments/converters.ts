import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

function convertToString(value?: object) {
  if (!value) {
    return undefined;
  }

  if (typeof value !== 'string') {
    return JSON.stringify(value);
  }

  return value;
}

export function getVolumeKey(deploymentId: string, volumeName: string) {
  return `${deploymentId}/${volumeName}`;
}

export function createVolumeEntity(deploymentUid: string, data: k8s.V1Volume) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.VOLUME._class,
        _type: Entities.VOLUME._type,
        _key: getVolumeKey(deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        'ephemeral.readOnly': data.ephemeral?.readOnly,
        'hostPath.path': data.hostPath?.path,
        'hostPath.type': data.hostPath?.type,
        classification: null,
        encrypted: false,
      },
    },
  });
}

export function getContainerSpecKey(deploymentId: string, specName: string) {
  return `${deploymentId}/${specName}`;
}

export function createContainerSpecEntity(
  deploymentUid: string,
  data: k8s.V1Container,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CONTAINER_SPEC._class,
        _type: Entities.CONTAINER_SPEC._type,
        _key: getContainerSpecKey(deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        image: data.image,
        imagePullPolicy: data.imagePullPolicy,
        'securityContext.allowPrivilegeEscalation':
          data.securityContext?.allowPrivilegeEscalation,
        'securityContext.privileged': data.securityContext?.privileged,
        'securityContext.procMount': data.securityContext?.procMount,
        'securityContext.readOnlyRootFilesystem':
          data.securityContext?.readOnlyRootFilesystem,
        'securityContext.runAsGroup': data.securityContext?.runAsGroup,
        'securityContext.runAsNonRoot': data.securityContext?.runAsNonRoot,
        'securityContext.runAsUser': data.securityContext?.runAsUser,
      },
    },
  });
}

export function createDeploymentEntity(data: k8s.V1Deployment) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.DEPLOYMENT._class,
        _type: Entities.DEPLOYMENT._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties
        minReadySeconds: data.spec?.minReadySeconds,
        paused: data.spec?.paused,
        progressDeadlineSeconds: data.spec?.progressDeadlineSeconds,
        replicas: data.spec?.replicas,
        revisionHistoryLimit: data.spec?.revisionHistoryLimit,
        'strategy.type': data.spec?.strategy?.type,
        'strategy.maxSurge': convertToString(
          data.spec?.strategy?.rollingUpdate?.maxSurge,
        ),
        'strategy.maxUnavailable': convertToString(
          data.spec?.strategy?.rollingUpdate?.maxUnavailable,
        ),
        // TODO: something about spec.template sections maybe?
        'status.availableReplicas': data.status?.availableReplicas,
        'status.collisionCount': data.status?.collisionCount,
        'status.observedGeneration': data.status?.observedGeneration,
        'status.readyReplicas': data.status?.readyReplicas,
        'status.replicas': data.status?.replicas,
        'status.unavailableReplicas': data.status?.unavailableReplicas,
        'status.updatedReplicas': data.status?.updatedReplicas,
      },
    },
  });
}
