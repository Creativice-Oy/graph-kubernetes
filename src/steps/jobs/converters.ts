import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createJobEntity(data: k8s.V1Job) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.JOB._class,
        _type: Entities.JOB._type,
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        deletedOn: parseTimePropertyValue(data.metadata?.deletionTimestamp),
        // spec fields
        activeDeadlineSeconds: data.spec?.activeDeadlineSeconds,
        backoffLimit: data.spec?.backoffLimit,
        completions: data.spec?.completions,
        manualSelector: data.spec?.manualSelector,
        parallelism: data.spec?.parallelism,
        ttlSecondsAfterFinished: data.spec?.ttlSecondsAfterFinished,
        'status.active': data.status?.active,
        'status.startTime': parseTimePropertyValue(data.status?.startTime),
        'status.completionTime': parseTimePropertyValue(
          data.status?.completionTime,
        ),
        'status.failed': data.status?.failed,
        'status.succeeded': data.status?.succeeded,
      },
    },
  });
}
