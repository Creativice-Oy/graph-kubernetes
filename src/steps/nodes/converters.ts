import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createNodeEntity(data: k8s.V1Node) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.NODE._class,
        _type: Entities.NODE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties - we could group them using "spec." notation
        podCIDR: data.spec?.podCIDR,
        podCIDRs: data.spec?.podCIDRs,
        providerID: data.spec?.providerID,
        unschedulable: data.spec?.unschedulable,
        'status.allocatable.cpu':
          data.status?.allocatable && data.status?.allocatable['cpu'],
        'status.allocatable.ephemeral-storage':
          data.status?.allocatable &&
          data.status?.allocatable['ephemeral-storage'],
        'status.allocatable.hugepages-1Gi':
          data.status?.allocatable && data.status?.allocatable['hugepages-1Gi'],
        'status.allocatable.hugepages-2Mi':
          data.status?.allocatable && data.status?.allocatable['hugepages-2Mi'],
        'status.allocatable.memory':
          data.status?.allocatable && data.status?.allocatable['memory'],
        'status.allocatable.pods':
          data.status?.allocatable && data.status?.allocatable['pods'],
        'status.capacity.cpu':
          data.status?.capacity && data.status?.capacity['cpu'],
        'status.capacity.ephemeral-storage':
          data.status?.capacity && data.status?.capacity['ephemeral-storage'],
        'status.capacity.hugepages-1Gi':
          data.status?.capacity && data.status?.capacity['hugepages-1Gi'],
        'status.capacity.hugepages-2Mi':
          data.status?.capacity && data.status?.capacity['hugepages-2Mi'],
        'status.capacity.memory':
          data.status?.capacity && data.status?.capacity['memory'],
        'status.capacity.pods':
          data.status?.capacity && data.status?.capacity['pods'],
        'status.kubeletEndpointPort':
          data.status?.daemonEndpoints?.kubeletEndpoint?.Port,
        'status.nodeInfo.architecture': data.status?.nodeInfo?.architecture,
        'status.nodeInfo.bootID': data.status?.nodeInfo?.bootID,
        'status.nodeInfo.containerRuntimeVersion':
          data.status?.nodeInfo?.containerRuntimeVersion,
        'status.nodeInfo.kernelVersion': data.status?.nodeInfo?.kernelVersion,
        'status.nodeInfo.kubeProxyVersion':
          data.status?.nodeInfo?.kubeProxyVersion,
        'status.nodeInfo.kubeletVersion': data.status?.nodeInfo?.kubeletVersion,
        'status.nodeInfo.machineID': data.status?.nodeInfo?.machineID,
        'status.nodeInfo.operatingSystem':
          data.status?.nodeInfo?.operatingSystem,
        'status.nodeInfo.osImage': data.status?.nodeInfo?.osImage,
        'status.nodeInfo.systemUUID': data.status?.nodeInfo?.systemUUID,
      },
    },
  });
}
