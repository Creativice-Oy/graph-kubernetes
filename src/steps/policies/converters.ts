import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createPodSecurityPolicyEntity(
  data: k8s.V1beta1PodSecurityPolicy,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.POD_SECURITY_POLICY._class,
        _type: Entities.POD_SECURITY_POLICY._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec
        // 5.2.5 Minimize the admission of containers with allowPrivilegeEscalation (Automated)
        allowPrivilegeEscalation: data.spec?.allowPrivilegeEscalation === true,
        // 5.2.8 Minimize the admission of containers with added capabilities (Automated)
        allowedCapabilities: data.spec?.allowedCapabilities,
        allowedProcMountTypes: data.spec?.allowedProcMountTypes,
        allowedUnsafeSysctls: data.spec?.allowedUnsafeSysctls,
        defaultAddCapabilities: data.spec?.defaultAddCapabilities,
        defaultAllowPrivilegeEscalation:
          data.spec?.defaultAllowPrivilegeEscalation,
        forbiddenSysctls: data.spec?.forbiddenSysctls,
        // 5.2.3 Minimize the admission of containers wishing to share the host IPC namespace (Automated)
        hostIPC: data.spec?.hostIPC === true,
        // 5.2.4 Minimize the admission of containers wishing to share the host network namespace (Automated)
        hostNetwork: data.spec?.hostNetwork === true,
        // 5.2.2 Minimize the admission of containers wishing to share the host process ID namespace (Automated)
        hostPID: data.spec?.hostPID === true,
        // 5.2.1 Minimize the admission of privileged containers (Automated)
        privileged: data.spec?.privileged === true,
        readOnlyRootFilesystem: data.spec?.readOnlyRootFilesystem === true,
        // 5.2.7 Minimize the admission of containers with the NET_RAW capability (Automated)
        // 5.2.9 Minimize the admission of containers with capabilities assigned (Manual)
        requiredDropCapabilities: data.spec?.requiredDropCapabilities,
        'runAsGroup.range': data.spec?.runAsGroup?.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        'runAsGroup.rule': data.spec?.runAsGroup?.rule,
        'runAsUser.range': data.spec?.runAsUser?.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        // 5.2.6 Minimize the admission of root containers (Automated)
        'runAsUser.rule': data.spec?.runAsUser?.rule,
        'runtimeClass.allowedRuntimeClassNames':
          data.spec?.runtimeClass?.allowedRuntimeClassNames,
        'runtimeClass.defaultRuntimeClassName':
          data.spec?.runtimeClass?.defaultRuntimeClassName,
        'seLinux.rule': data.spec?.seLinux.rule,
        'seLinux.seLinuxOptions.level':
          data.spec?.seLinux.seLinuxOptions?.level,
        'seLinux.seLinuxOptions.role': data.spec?.seLinux.seLinuxOptions?.role,
        'seLinux.seLinuxOptions.type': data.spec?.seLinux.seLinuxOptions?.type,
        'seLinux.seLinuxOptions.user': data.spec?.seLinux.seLinuxOptions?.user,
        'supplementalGroups.range': data.spec?.supplementalGroups.ranges?.map(
          (range) => `${range.min}, ${range.max}`,
        ),
        'supplementalGroups.rule': data.spec?.supplementalGroups?.rule,
      },
    },
  });
}

export function createNetworkPolicyEntity(data: k8s.V1NetworkPolicy) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.NETWORK_POLICY._class,
        _type: Entities.NETWORK_POLICY._type,
        // metadata properties
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec
        // TODO: get egress data (mapping? separate entities? combination?)
        // TODO: get ingress data (mapping? separate entities? combination?)
        policyTypes: data.spec?.policyTypes,
      },
    },
  });
}
