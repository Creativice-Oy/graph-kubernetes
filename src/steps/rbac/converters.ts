import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createRoleEntity(data: k8s.V1Role) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ROLE._class,
        _type: Entities.ROLE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        namespace: data.metadata?.namespace,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createClusterRoleEntity(data: k8s.V1ClusterRole) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CLUSTER_ROLE._class,
        _type: Entities.CLUSTER_ROLE._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createRoleBindingEntity(data: k8s.V1RoleBinding) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ROLE_BINDING._class,
        _type: Entities.ROLE_BINDING._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createClusterRoleBindingEntity(data: k8s.V1ClusterRoleBinding) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CLUSTER_ROLE_BINDING._class,
        _type: Entities.CLUSTER_ROLE_BINDING._type,
        _key: data.metadata?.uid,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}
