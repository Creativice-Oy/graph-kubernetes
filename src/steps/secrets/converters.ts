import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createSecretEntity(configMap: k8s.V1Secret) {
  // Secret contains a data which we don't want to ingest.
  const { data, ...withoutData } = configMap;

  return createIntegrationEntity({
    entityData: {
      source: withoutData,
      assign: {
        _class: Entities.SECRET._class,
        _type: Entities.SECRET._type,
        _key: withoutData.metadata!.uid!,
        name: withoutData.metadata?.name,
        displayName: withoutData.metadata?.name,
        generation: withoutData.metadata?.generation,
        deletionGracePeriodSeconds:
          withoutData.metadata?.deletionGracePeriodSeconds,
        resourceVersion: withoutData.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(
          withoutData.metadata?.creationTimestamp,
        ),
        deletedOn: parseTimePropertyValue(
          withoutData.metadata?.deletionTimestamp,
        ),
        type: withoutData.type,
        classification: null,
        encrypted: false,
      },
    },
  });
}
