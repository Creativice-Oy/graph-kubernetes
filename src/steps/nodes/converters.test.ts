import { createNodeEntity } from './converters';
import { createMockNode } from '../../../test/mocks';

describe('#createNodeEntity', () => {
  test('should convert data', () => {
    expect(createNodeEntity(createMockNode())).toMatchSnapshot();
  });

  test('should parse status.addresses correctly', () => {
    expect(
      createNodeEntity({
        ...createMockNode(),
        status: {
          addresses: [
            {
              address: '1.1.1.1',
              type: 'InternalIP',
            },
            {
              address: '2.2.2.2',
              type: 'InternalIP',
            },
            {
              address: '5.5.5.5',
              type: 'ExternalIP',
            },
            {
              address: '6.6.6.6',
              type: 'ExternalIP',
            },
            {
              address: 'minikube',
              type: 'Hostname',
            },
            {
              address: 'internalDns',
              type: 'InternalDns',
            },
            {
              address: 'externalDns',
              type: 'ExternalDns',
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });
});
