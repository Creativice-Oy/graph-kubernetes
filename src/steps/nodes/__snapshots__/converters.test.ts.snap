// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createNodeEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Group",
  ],
  "_key": "103b35d1-8951-455c-95cb-3da4af29bd71",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "metadata": Object {
          "annotations": Object {
            "kubeadm.alpha.kubernetes.io/cri-socket": "/var/run/dockershim.sock",
            "node.alpha.kubernetes.io/ttl": "0",
            "volumes.kubernetes.io/controller-managed-attach-detach": "true",
          },
          "creationTimestamp": 2021-04-13T15:20:38.000Z,
          "labels": Object {
            "beta.kubernetes.io/arch": "amd64",
            "beta.kubernetes.io/os": "linux",
            "kubernetes.io/arch": "amd64",
            "kubernetes.io/hostname": "minikube",
            "kubernetes.io/os": "linux",
            "minikube.k8s.io/commit": "15cede53bdc5fe242228853e737333b09d4336b5",
            "minikube.k8s.io/name": "minikube",
            "minikube.k8s.io/updated_at": "2021_04_13T17_20_41_0700",
            "minikube.k8s.io/version": "v1.19.0",
            "node-role.kubernetes.io/control-plane": "",
            "node-role.kubernetes.io/master": "",
          },
          "managedFields": Array [
            Object {
              "apiVersion": "v1",
              "fieldsType": "FieldsV1",
              "fieldsV1": Object {
                "f:metadata": Object {
                  "f:annotations": Object {
                    ".": Object {},
                    "f:volumes.kubernetes.io/controller-managed-attach-detach": Object {},
                  },
                  "f:labels": Object {
                    ".": Object {},
                    "f:beta.kubernetes.io/arch": Object {},
                    "f:beta.kubernetes.io/os": Object {},
                    "f:kubernetes.io/arch": Object {},
                    "f:kubernetes.io/hostname": Object {},
                    "f:kubernetes.io/os": Object {},
                  },
                },
                "f:status": Object {
                  "f:addresses": Object {
                    ".": Object {},
                    "k:{\\"type\\":\\"Hostname\\"}": Object {
                      ".": Object {},
                      "f:address": Object {},
                      "f:type": Object {},
                    },
                    "k:{\\"type\\":\\"InternalIP\\"}": Object {
                      ".": Object {},
                      "f:address": Object {},
                      "f:type": Object {},
                    },
                  },
                  "f:allocatable": Object {
                    ".": Object {},
                    "f:cpu": Object {},
                    "f:ephemeral-storage": Object {},
                    "f:hugepages-1Gi": Object {},
                    "f:hugepages-2Mi": Object {},
                    "f:memory": Object {},
                    "f:pods": Object {},
                  },
                  "f:capacity": Object {
                    ".": Object {},
                    "f:cpu": Object {},
                    "f:ephemeral-storage": Object {},
                    "f:hugepages-1Gi": Object {},
                    "f:hugepages-2Mi": Object {},
                    "f:memory": Object {},
                    "f:pods": Object {},
                  },
                  "f:conditions": Object {
                    ".": Object {},
                    "k:{\\"type\\":\\"DiskPressure\\"}": Object {
                      ".": Object {},
                      "f:lastHeartbeatTime": Object {},
                      "f:lastTransitionTime": Object {},
                      "f:message": Object {},
                      "f:reason": Object {},
                      "f:status": Object {},
                      "f:type": Object {},
                    },
                    "k:{\\"type\\":\\"MemoryPressure\\"}": Object {
                      ".": Object {},
                      "f:lastHeartbeatTime": Object {},
                      "f:lastTransitionTime": Object {},
                      "f:message": Object {},
                      "f:reason": Object {},
                      "f:status": Object {},
                      "f:type": Object {},
                    },
                    "k:{\\"type\\":\\"PIDPressure\\"}": Object {
                      ".": Object {},
                      "f:lastHeartbeatTime": Object {},
                      "f:lastTransitionTime": Object {},
                      "f:message": Object {},
                      "f:reason": Object {},
                      "f:status": Object {},
                      "f:type": Object {},
                    },
                    "k:{\\"type\\":\\"Ready\\"}": Object {
                      ".": Object {},
                      "f:lastHeartbeatTime": Object {},
                      "f:lastTransitionTime": Object {},
                      "f:message": Object {},
                      "f:reason": Object {},
                      "f:status": Object {},
                      "f:type": Object {},
                    },
                  },
                  "f:daemonEndpoints": Object {
                    "f:kubeletEndpoint": Object {
                      "f:Port": Object {},
                    },
                  },
                  "f:images": Object {},
                  "f:nodeInfo": Object {
                    "f:architecture": Object {},
                    "f:bootID": Object {},
                    "f:containerRuntimeVersion": Object {},
                    "f:kernelVersion": Object {},
                    "f:kubeProxyVersion": Object {},
                    "f:kubeletVersion": Object {},
                    "f:machineID": Object {},
                    "f:operatingSystem": Object {},
                    "f:osImage": Object {},
                    "f:systemUUID": Object {},
                  },
                },
              },
              "manager": "kubelet",
              "operation": "Update",
              "time": 2021-04-13T15:20:38.000Z,
            },
            Object {
              "apiVersion": "v1",
              "fieldsType": "FieldsV1",
              "fieldsV1": Object {
                "f:metadata": Object {
                  "f:annotations": Object {
                    "f:kubeadm.alpha.kubernetes.io/cri-socket": Object {},
                  },
                  "f:labels": Object {
                    "f:node-role.kubernetes.io/control-plane": Object {},
                    "f:node-role.kubernetes.io/master": Object {},
                  },
                },
              },
              "manager": "kubeadm",
              "operation": "Update",
              "time": 2021-04-13T15:20:41.000Z,
            },
            Object {
              "apiVersion": "v1",
              "fieldsType": "FieldsV1",
              "fieldsV1": Object {
                "f:metadata": Object {
                  "f:labels": Object {
                    "f:minikube.k8s.io/commit": Object {},
                    "f:minikube.k8s.io/name": Object {},
                    "f:minikube.k8s.io/updated_at": Object {},
                    "f:minikube.k8s.io/version": Object {},
                  },
                },
              },
              "manager": "kubectl-label",
              "operation": "Update",
              "time": 2021-04-13T15:20:42.000Z,
            },
            Object {
              "apiVersion": "v1",
              "fieldsType": "FieldsV1",
              "fieldsV1": Object {
                "f:metadata": Object {
                  "f:annotations": Object {
                    "f:node.alpha.kubernetes.io/ttl": Object {},
                  },
                },
                "f:spec": Object {
                  "f:podCIDR": Object {},
                  "f:podCIDRs": Object {
                    ".": Object {},
                    "v:\\"10.10.10.10/24\\"": Object {},
                  },
                },
              },
              "manager": "kube-controller-manager",
              "operation": "Update",
              "time": 2021-04-13T15:20:56.000Z,
            },
          ],
          "name": "minikube",
          "resourceVersion": "81209",
          "uid": "103b35d1-8951-455c-95cb-3da4af29bd71",
        },
        "spec": Object {
          "podCIDR": "10.10.10.10/24",
          "podCIDRs": Array [
            "10.10.10.10/24",
          ],
        },
        "status": Object {
          "addresses": Array [
            Object {
              "address": "111.111.11.1",
              "type": "InternalIP",
            },
            Object {
              "address": "minikube",
              "type": "Hostname",
            },
          ],
          "allocatable": Object {
            "cpu": "8",
            "ephemeral-storage": "88103912Ki",
            "hugepages-1Gi": "0",
            "hugepages-2Mi": "0",
            "memory": "16289508Ki",
            "pods": "110",
          },
          "capacity": Object {
            "cpu": "8",
            "ephemeral-storage": "88103912Ki",
            "hugepages-1Gi": "0",
            "hugepages-2Mi": "0",
            "memory": "16289508Ki",
            "pods": "110",
          },
          "conditions": Array [
            Object {
              "lastHeartbeatTime": 2021-04-15T16:44:09.000Z,
              "lastTransitionTime": 2021-04-13T15:20:33.000Z,
              "message": "kubelet has sufficient memory available",
              "reason": "KubeletHasSufficientMemory",
              "status": "False",
              "type": "MemoryPressure",
            },
          ],
          "daemonEndpoints": Object {
            "kubeletEndpoint": Object {
              "Port": 10250,
            },
          },
          "images": Array [
            Object {
              "names": Array [
                "k8s.gcr.io/etcd@sha256:4ad90a11b55313b182afc186b9876c8e891531b8db4c9bf1541953021618d0e2",
                "k8s.gcr.io/etcd:3.4.13-0",
              ],
              "sizeBytes": 253392289,
            },
          ],
          "nodeInfo": Object {
            "architecture": "amd64",
            "bootID": "a711daa9-dbe2-4923-916c-b8693ced170d",
            "containerRuntimeVersion": "docker://20.10.5",
            "kernelVersion": "5.8.0-49-generic",
            "kubeProxyVersion": "v1.20.2",
            "kubeletVersion": "v1.20.2",
            "machineID": "73c9fceff7724090ba72b64bf6e8eff8",
            "operatingSystem": "linux",
            "osImage": "Ubuntu 20.04.1 LTS",
            "systemUUID": "444a01a7-d999-4a38-a4d9-faaee5ee05f5",
          },
        },
      },
    },
  ],
  "_type": "kube_node",
  "active": undefined,
  "createdOn": 1618327238000,
  "deletionGracePeriodSeconds": undefined,
  "displayName": "minikube",
  "generation": undefined,
  "name": "minikube",
  "podCIDR": "10.10.10.10/24",
  "podCIDRs": Array [
    "10.10.10.10/24",
  ],
  "providerID": undefined,
  "resourceVersion": "81209",
  "status.allocatable.cpu": "8",
  "status.allocatable.ephemeral-storage": "88103912Ki",
  "status.allocatable.hugepages-1Gi": "0",
  "status.allocatable.hugepages-2Mi": "0",
  "status.allocatable.memory": "16289508Ki",
  "status.allocatable.pods": "110",
  "status.capacity.cpu": "8",
  "status.capacity.ephemeral-storage": "88103912Ki",
  "status.capacity.hugepages-1Gi": "0",
  "status.capacity.hugepages-2Mi": "0",
  "status.capacity.memory": "16289508Ki",
  "status.capacity.pods": "110",
  "status.kubeletEndpointPort": 10250,
  "status.nodeInfo.architecture": "amd64",
  "status.nodeInfo.bootID": "a711daa9-dbe2-4923-916c-b8693ced170d",
  "status.nodeInfo.containerRuntimeVersion": "docker://20.10.5",
  "status.nodeInfo.kernelVersion": "5.8.0-49-generic",
  "status.nodeInfo.kubeProxyVersion": "v1.20.2",
  "status.nodeInfo.kubeletVersion": "v1.20.2",
  "status.nodeInfo.machineID": "73c9fceff7724090ba72b64bf6e8eff8",
  "status.nodeInfo.operatingSystem": "linux",
  "status.nodeInfo.osImage": "Ubuntu 20.04.1 LTS",
  "status.nodeInfo.systemUUID": "444a01a7-d999-4a38-a4d9-faaee5ee05f5",
  "unschedulable": undefined,
}
`;
